src/main/java/net/starlark/java/eval/StarlarkInt.java

```diff
@@ -505,10 +505,32 @@ public static StarlarkInt subtract(StarlarkInt x, StarlarkInt y) {
 
   /** Returns x * y. */
   public static StarlarkInt multiply(StarlarkInt x, StarlarkInt y) {
-    if (x instanceof Int32 && y instanceof Int32) {
-      long xl = ((Int32) x).v;
-      long yl = ((Int32) y).v;
-      return StarlarkInt.of(xl * yl);
+    try {
+      long xl = x.toLongFast();
+      long yl = y.toLongFast();
+
+      // Copy-paste from Guava LongMath.saturatedMultiply https://git.io/JIlZP
+
+      long zl = xl * yl;
+
+      int leadingZeros =
+        Long.numberOfLeadingZeros(xl)
+          + Long.numberOfLeadingZeros(~xl)
+          + Long.numberOfLeadingZeros(yl)
+          + Long.numberOfLeadingZeros(~yl);
+
+      if (leadingZeros > Long.SIZE + 1) {
+        return StarlarkInt.of(zl);
+      }
+      if (leadingZeros < Long.SIZE || (xl < 0 && yl == Long.MIN_VALUE)) {
+        throw OVERFLOW;
+      }
+      if (xl == 0 || zl / xl == yl) {
```

Comment:
There are better ways to do this without using division. See fig 8.2 of Hacker's Delight, for example.  Or call Long.multiplyHigh, which uses exactly that algorithm.

Reply:
`Math.multiplyHigh` was added in Java 9, so it cannot be used (and I did not know such thing exists). Let me think...

Reply:
> Math.multiplyHigh was added in Java 9, so it cannot be used (and I did not know such thing exists). Let me think...

It uses an algorithm from HD, so we can make a local copy of the code from the book.